### PR TITLE
Support Ray Runtime in init_orca_context

### PR DESCRIPTION
**Pull request by sgwhat**
- Update init_orca_context, for now, init_orca_context support "ray" or "spark" backend. 

> spark backend is the default runtime, which means you could call init_orca_context as usual. 

> ray backend is the new runtime that we support this time, if you wanna create a pure raycontext, you should call `init_orca_context(runtime="ray", address="localhost:6379")`, the redis_address is required.

- Convert the original raycontext to ray_on_spark_context, the new raycontext support pure ray runtime and ray_on_spark.
- Add TF/Pytorch estimator and RayContext unit tests.

### CODE REVIEW

**Comment by jason-dai** [comment]
See my comments at https://github.com/intel-analytics/BigDL/pull/3640

File: python/orca/src/bigdl/orca/common.py [hunk 1]
```diff
@@ -167,12 +167,14 @@ def _check_python_micro_version():
                            f"with micro version >= 10 (e.g. 3.{sys.version_info[1]}.10)")
 
 
-def init_orca_context(cluster_mode=None, cores=2, memory="2g", num_nodes=1,
+def init_orca_context(runtime="spark", cluster_mode=None, cores=2, memory="2g", num_nodes=1,
```
Comment (jason-dai) [inline python/orca/src/bigdl/orca/common.py:170]: change to `cluster_mode=None, runtime="spark"`

File: python/orca/src/bigdl/orca/common.py [hunk 2]
```diff
@@ -167,12 +167,14 @@ def _check_python_micro_version():
...
                       init_ray_on_spark=False, **kwargs):
     """
     Creates or gets a SparkContext for different Spark cluster modes (and launch Ray services
     across the cluster if necessary).
 
+    :param runtime: The runtime for backend. One of "ray", "ray_on_spark" and "spark",
```
Comment (jason-dai) [inline python/orca/src/bigdl/orca/common.py:176]: Change to `"spark" and "ray"`

In future, we may want to support "spark_and_ray"

File: python/orca/src/bigdl/orca/common.py [hunk 3]
```diff
@@ -206,105 +208,117 @@ def init_orca_context(cluster_mode=None, cores=2, memory="2g", num_nodes=1,
...
-                                   num_executors=num_nodes, executor_cores=cores,
-                                   executor_memory=memory, **spark_args)
-        elif cluster_mode == "standalone":
-            for key in ["driver_cores", "driver_memory", "extra_executor_memory_for_ray",
-                        "extra_python_lib", "jars", "master", "python_location",
-                        "enable_numa_binding"]:
+                raise ValueError("cluster_mode can only be local, yarn-client, yarn-cluster,"
+                                 "k8s-client or standalone, "
+                                 "but got: %s".format(cluster_mode))
+        if runtime == "ray_on_spark":
```
Comment (jason-dai) [inline python/orca/src/bigdl/orca/common.py:307]: remove this one

File: python/orca/src/bigdl/orca/common.py [hunk 4]
```diff
@@ -206,105 +208,117 @@ def init_orca_context(cluster_mode=None, cores=2, memory="2g", num_nodes=1,
...
-        elif cluster_mode == "spark-submit":
+    if runtime == "ray":
+        assert cluster_mode is None, "cluster mode should be None"
+        from bigdl.orca.ray import RayContext
+        ray_ctx = RayContext(runtime="ray", cores=cores, num_nodes=num_nodes,
+                             **kwargs)
+        assert "address" in kwargs, "ray_address must be specified if runtime is ray"
+        ray_ctx.init()
+        return ray_ctx
+    elif runtime in ("spark", "ray_on_spark"):
```
Comment (jason-dai) [inline python/orca/src/bigdl/orca/common.py:219]: `ray_on_spark` is not supported; we should continue to use `init_ray_on_spark`

File: python/orca/src/bigdl/orca/common.py [hunk 5]
```diff
@@ -317,7 +315,7 @@ def init_orca_context(runtime="spark", cluster_mode=None, cores=2, memory="2g",
                 driver_cores = 0  # This is the default value.
                 ray_ctx.init(driver_cores=driver_cores)
     else:
-        raise ValueError("runtime can only be ray, spark or ray_on_spark, "
+        raise ValueError("runtime can only be ray and spark, "
```
Comment (jason-dai) [inline python/orca/src/bigdl/orca/common.py:318]: "spark" or "ray"

File: python/orca/src/bigdl/orca/common.py [hunk 6]
```diff
@@ -328,3 +339,9 @@ def stop_orca_context():
             from bigdl.dllib.nncontext import stop_spark_standalone
             stop_spark_standalone()
         sc.stop()
+    else:
+        if RayContext._active_ray_context is not None:
+            print("Stopping ray_orca context")
+            ray_ctx = RayContext.get(init_orca_context)
```
Comment (jason-dai) [inline python/orca/src/bigdl/orca/common.py:345]: why `init_orca_context` here?

File: python/orca/src/bigdl/orca/ray/raycontext.py [hunk 7]
```diff
@@ -497,149 +67,12 @@ def stop(self):
...
-        ray.shutdown()
-        init_params = dict(
-            address=redis_address,
-            _redis_password=self.ray_service.password,
-            _node_ip_address=node_ip
-        )
-        if version.parse(ray.__version__) >= version.parse("1.4.0"):
-            init_params["namespace"] = "az"
-        return ray.init(**init_params)
+            raise Exception("No active RayContext. Please create a RayContext and init it first")
```
Comment (jason-dai) [inline python/orca/src/bigdl/orca/ray/raycontext.py:78]: The user won't know how to create `RayContext`; I think you should just ask the user to call `init_orca_context` first.

**Comment by sgwhat** [comment]
Jenkins Passed: http://10.112.231.51:18888/job/BigDL-PRVN-ray-ctx/28/

File: python/orca/src/bigdl/orca/common.py [hunk 8]
```diff
@@ -206,120 +207,130 @@ def init_orca_context(cluster_mode=None, cores=2, memory="2g", num_nodes=1,
...
-                warnings.warn("For Spark local mode, num_nodes should be 1, but got "
-                              + repr(num_nodes) + ", ignored", Warning)
-            os.environ["SPARK_DRIVER_MEMORY"] = memory
-            if "python_location" in kwargs:
-                spark_args["python_location"] = kwargs["python_location"]
-            from bigdl.dllib.nncontext import init_spark_on_local
-            sc = init_spark_on_local(cores, **spark_args)
-        elif cluster_mode == "spark-submit":
+    if runtime == "ray":
+        assert cluster_mode is None, "cluster mode should be None"
```
Comment (yangw1234) [inline python/orca/src/bigdl/orca/common.py:211]: Maybe a more clear message? "Currently, cluster mode is not supported for ray runtime and you must connect to an exiting ray cluster."

File: python/orca/test/bigdl/orca/ray/ray_cluster/ray_est/test_ray_pytorch_estimator.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        return y
+
+def get_model(config):
+    torch.manual_seed(0)
+    return Net()
+
+def get_optimizer(model, config):
+    return torch.optim.SGD(model.parameters(), lr=config.get("lr", 1e-2))
+
+if __name__ == "__main__":
```
Comment (yangw1234) [inline python/orca/test/bigdl/orca/ray/ray_cluster/ray_est/test_ray_pytorch_estimator.py:84]: Does it work for pytest?

Comment (yangw1234) [reply python/orca/test/bigdl/orca/ray/ray_cluster/ray_est/test_ray_pytorch_estimator.py:84]: I think test case should write in a function prefixed with "test_"

File: python/orca/test/bigdl/orca/ray/ray_cluster/ray_est/test_ray_tf2estimator.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        "metrics": ["mean_squared_error"]
+    }
+    return args
+
+def model_creator(config):
+    model = simple_model(config)
+    model.compile(**compile_args(config))
+    return model
+
+if __name__ == "__main__":
```
Comment (yangw1234) [inline python/orca/test/bigdl/orca/ray/ray_cluster/ray_est/test_ray_tf2estimator.py:73]: Wrap the following code in a test case (function or class)?

**Comment by yangw1234** [comment]
By the way, I think it is better to use a more meaningful and specific PR name , as well as a short high level description of the PR explaining the changes you made.

**Comment by sgwhat** [comment]
> By the way, I think it is better to use a more meaningful and specific PR name , as well as a short high level description of the PR explaining the changes you made.

How about Support ray runtime in init_orca_context and raycontext ?

**Comment by sgwhat** [comment]
Jenkins retest passed: http://10.112.231.51:18888/job/BigDL-PRVN-ray-ctx/47/

**Review by yangw1234 — APPROVED** [review]
LGTM

**Pull request merged**